[BUG] dpp tests fail

**Describe the bug**
Our IT 22.02 Dataproc tests failed, it also failed on our integration nightly tests for 3.1.1.

```
07:00:43  FAILED integration_tests/src/main/python/dpp_test.py::test_dpp_reuse_broadcast_exchange_aqe_off[0-parquet][IGNORE_ORDER]
07:00:43  FAILED integration_tests/src/main/python/dpp_test.py::test_dpp_reuse_broadcast_exchange_aqe_off[0-orc][IGNORE_ORDER]
07:00:43  FAILED integration_tests/src/main/python/dpp_test.py::test_dpp_reuse_broadcast_exchange_aqe_off[1-parquet][IGNORE_ORDER]
07:00:43  FAILED integration_tests/src/main/python/dpp_test.py::test_dpp_reuse_broadcast_exchange_aqe_off[1-orc][IGNORE_ORDER]
07:00:43  FAILED integration_tests/src/main/python/dpp_test.py::test_dpp_reuse_broadcast_exchange_cpu_scan[IGNORE_ORDER, ALLOW_NON_GPU(FileSourceScanExec)]
```

one of the stack traces:

```
07:00:43  E                   py4j.protocol.Py4JJavaError: An error occurred while calling z:com.nvidia.spark.rapids.ExecutionPlanCaptureCallback.assertContains.
07:00:43  E                   : java.lang.AssertionError: assertion failed: Could not find GpuSubqueryBroadcastExec in the Spark plan
07:00:43  E                   GpuColumnarToRow false
07:00:43  E                   +- GpuSort [key#988 ASC NULLS FIRST, sum(value)#1002L ASC NULLS FIRST], true, com.nvidia.spark.rapids.OutOfCoreSort$@40b3567f
07:00:43  E                      +- GpuShuffleCoalesce 2147483647
07:00:43  E                         +- GpuColumnarExchange gpurangepartitioning(key#988 ASC NULLS FIRST, sum(value)#1002L ASC NULLS FIRST, 48), ENSURE_REQUIREMENTS, [id=#2233]
07:00:43  E                            +- GpuHashAggregate(keys=[key#988], functions=[gpusum(cast(value#987 as bigint), LongType, false, false)], output=[key#988, sum(value)#1002L])
07:00:43  E                               +- GpuShuffleCoalesce 2147483647
07:00:43  E                                  +- GpuColumnarExchange gpuhashpartitioning(key#988, 48), ENSURE_REQUIREMENTS, [id=#2225]
07:00:43  E                                     +- GpuHashAggregate(keys=[key#988], functions=[partial_gpusum(cast(value#987 as bigint), LongType, false, false)], output=[key#988, sum#1007L])
07:00:43  E                                        +- GpuProject [value#987, key#988]
07:00:43  E                                           +- GpuBroadcastHashJoin [key#988], [key#989], Inner, GpuBuildRight
07:00:43  E                                              :- GpuCoalesceBatches targetsize(2147483647)
07:00:43  E                                              :  +- GpuFilter (gpuisnotnull(value#987) AND (value#987 > 0)), true
07:00:43  E                                              :     +- HostColumnarToGpu targetsize(2147483647)
07:00:43  E                                              :        +- FileScan parquet default.tmp_table_427264_0[value#987,key#988] Batched: true, DataFilters: [isnotnull(value#987), (value#987 > 0)], Format: Parquet, Location: InMemoryFileIndex[hdfs://rapids-it-383-m/user/hive/warehouse/tmp_table_427264_0/key=0, hdfs://rap..., PartitionFilters: [isnotnull(key#988), dynamicpruningexpression(key#988 IN dynamicpruning#1005)], PushedFilters: [IsNotNull(value), GreaterThan(value,0)], ReadSchema: struct<value:int>
07:00:43  E                                              :              +- SubqueryBroadcast dynamicpruning#1005, 0, [key#989], [id=#2140]
07:00:43  E                                              :                 +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)),false), [id=#2139]
07:00:43  E                                              :                    +- GpuColumnarToRow false
07:00:43  E                                              :                       +- GpuProject [key#989]
07:00:43  E                                              :                          +- GpuCoalesceBatches targetsize(2147483647)
07:00:43  E                                              :                             +- GpuFilter ((gpuisnotnull(filter#991) AND (filter#991 = 1552)) AND gpuisnotnull(key#989)), true
07:00:43  E                                              :                                +- HostColumnarToGpu targetsize(2147483647)
07:00:43  E                                              :                                   +- FileScan parquet default.tmp_table_427264_1[key#989,filter#991] Batched: true, DataFilters: [isnotnull(filter#991), (filter#991 = 1552), isnotnull(key#989)], Format: Parquet, Location: InMemoryFileIndex[hdfs://rapids-it-383-m/user/hive/warehouse/tmp_table_427264_1], PartitionFilters: [], PushedFilters: [IsNotNull(filter), EqualTo(filter,1552), IsNotNull(key)], ReadSchema: struct<key:int,filter:int>
07:00:43  E                                              +- GpuBroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)),false), [id=#2221]
07:00:43  E                                                 +- GpuProject [key#989]
07:00:43  E                                                    +- GpuCoalesceBatches targetsize(2147483647)
07:00:43  E                                                       +- GpuFilter ((gpuisnotnull(filter#991) AND (filter#991 = 1552)) AND gpuisnotnull(key#989)), true
07:00:43  E                                                          +- HostColumnarToGpu targetsize(2147483647)
07:00:43  E                                                             +- FileScan parquet default.tmp_table_427264_1[key#989,filter#991] Batched: true, DataFilters: [isnotnull(filter#991), (filter#991 = 1552), isnotnull(key#989)], Format: Parquet, Location: InMemoryFileIndex[hdfs://rapids-it-383-m/user/hive/warehouse/tmp_table_427264_1], PartitionFilters: [], PushedFilters: [IsNotNull(filter), EqualTo(filter,1552), IsNotNull(key)], ReadSchema: struct<key:int,filter:int>
07:00:43  E                   
07:00:43  E                   	at scala.Predef$.assert(Predef.scala:223)
```

--- a/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
+++ b/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
@@ -363,6 +363,25 @@ abstract class Spark31XShims extends Spark301until320Shims with Logging {
         ExecChecks((TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.STRUCT + TypeSig.MAP +
             TypeSig.ARRAY + TypeSig.DECIMAL_128_FULL).nested(), TypeSig.all),
         (fsse, conf, p, r) => new SparkPlanMeta[FileSourceScanExec](fsse, conf, p, r) {
+
+          // Replaces SubqueryBroadcastExec inside dynamic pruning filters with GPU counterpart
+          // if possible. Instead regarding filters as childExprs of current Meta, we create
+          // a new meta for SubqueryBroadcastExec. The reason is that the GPU replacement of
+          // FileSourceScan is independent from the replacement of the partitionFilters. It is
+          // possible that the FileSourceScan is on the CPU, while the dynamic partitionFilters
+          // are on the GPU. And vice versa.
+          private lazy val partitionFilters = wrapped.partitionFilters.map { filter =>
+            filter.transformDown {
+              case dpe @ DynamicPruningExpression(inSub: InSubqueryExec)
+                if inSub.plan.isInstanceOf[SubqueryBroadcastExec] =>
+
+                val subBcMeta = GpuOverrides.wrapAndTagPlan(inSub.plan, conf)
+                subBcMeta.tagForExplain()
+                val gpuSubBroadcast = subBcMeta.convertIfNeeded().asInstanceOf[BaseSubqueryExec]
+                dpe.copy(inSub.copy(plan = gpuSubBroadcast))
+            }
+          }
+
           // partition filters and data filters are not run on the GPU
           override val childExprs: Seq[ExprMeta[_]] = Seq.empty
 

--- a/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
+++ b/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
@@ -387,6 +387,10 @@ abstract class Spark31XShims extends Spark301until320Shims with Logging {
 
           override def tagPlanForGpu(): Unit = GpuFileSourceScanExec.tagSupport(this)
 
+          override def convertToCpu(): SparkPlan = {
+            wrapped.copy(partitionFilters = partitionFilters)
+          }
+
           override def convertToGpu(): GpuExec = {
             val sparkSession = wrapped.relation.sparkSession
             val options = wrapped.relation.options
@@ -394,7 +398,7 @@ abstract class Spark31XShims extends Spark301until320Shims with Logging {
             val location = replaceWithAlluxioPathIfNeeded(
               conf,
               wrapped.relation,
-              wrapped.partitionFilters,
+              partitionFilters,
               wrapped.dataFilters)
 
             val newRelation = HadoopFsRelation(
@@ -409,9 +413,10 @@ abstract class Spark31XShims extends Spark301until320Shims with Logging {
               newRelation,
               wrapped.output,
               wrapped.requiredSchema,
-              wrapped.partitionFilters,
+              partitionFilters,
               wrapped.optionalBucketSet,
-              wrapped.optionalNumCoalescedBuckets,
+              // TODO: Does Databricks have coalesced bucketing implemented?
+              None,
               wrapped.dataFilters,
               wrapped.tableIdentifier)(conf)
           }

--- a/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
+++ b/sql-plugin/src/main/311until320-nondb/scala/com/nvidia/spark/rapids/shims/v2/Spark31XShims.scala
@@ -415,8 +415,7 @@ abstract class Spark31XShims extends Spark301until320Shims with Logging {
               wrapped.requiredSchema,
               partitionFilters,
               wrapped.optionalBucketSet,
-              // TODO: Does Databricks have coalesced bucketing implemented?
-              None,
+              wrapped.optionalNumCoalescedBuckets,
               wrapped.dataFilters,
               wrapped.tableIdentifier)(conf)
           }

